google/cloud/spanner_dbapi/cursor.py

```diff
@@ -258,9 +260,56 @@ def executemany(self, operation, seq_of_params):
 
         many_result_set = StreamedManyResultSets()
 
-        for params in seq_of_params:
-            self.execute(operation, params)
-            many_result_set.add_iter(self._itr)
+        if classification in (parse_utils.STMT_INSERT, parse_utils.STMT_UPDATING):
+            statements = []
+
+            for params in seq_of_params:
+                sql, params = parse_utils.sql_pyformat_args_to_spanner(
+                    operation, params
+                )
+                statements.append((sql, params, get_param_types(params)))
+
+            if self.connection.autocommit:
+                transaction = self.connection.transaction_checkout()
+                status, res = transaction.batch_update(statements)
+                many_result_set.add_iter(res)
+
+                if status.code != OK:
+                    self.connection._transaction = None
+                    raise OperationalError(status.details)
+
+                transaction.commit()
+            else:
+                retried = False
+                while True:
+                    try:
+                        transaction = self.connection.transaction_checkout()
+
+                        res_checksum = ResultsChecksum()
+                        if not retried:
+                            self.connection._statements.append(
+                                (statements, res_checksum)
+                            )
+
+                        status, res = transaction.batch_update(statements)
+                        many_result_set.add_iter(res)
+                        res_checksum.consume_result(res)
```

Comment:
The status should be added to the checksum as well. Consider the following situation:
1. During the initial attempt batch_update returns [1, 1, 1], Status.INVALID_ARGUMENT
2. During a retry attempt batch_update returns [1, 1, 1], Status.ALREADY_EXISTS

These two results should not be considered equal and the retry should be aborted.

Reply:
Hm-m, yeah, it makes sense. Done